### DISCUSSION TITLE
egolt start lsp with only specific module from a repo -- project.el question

### DISCUSSION THREAD

**Discussion by saladmaker** [joaotavora/eglot#1336] [Q&A] [answered]
**Eglot** relies on **project package** which has vs-aware behavior and sets the project root to the repo, but this causes lsp server to hangs indefinitely for multimodule (100>) projects
How to configure eglot to launch lsp only for some module.
I'm using the java **netbeans-lsp** i encountered the same problem with **lsp-mode** and **jdtls**.

**Comment by joaotavora** [comment, marked as answer]
I'm having a hard time understanding your request, but it sounds familiar.  You have two options:

1. You tell your Emacs session as a whoel that the root of the project is not the root of the repo. -- for that particular project.  That affects everything project related, not just Eglot.

2. You tell Eglot that the project it should consider as the root of the LSP project is not what `project.el` by defaults tells it.

As far as I know, you can do both things with `project.el` 's interfaces.  Maybe there's an easy way to do the first.  Ask the project.el manager @dgutov .

For the second option, you have to use Elisp, adding a member function to `project-find-functions`, the documentation of which you should read carefully.  That new function can inspect the variable `eglot-lsp-context` with `bound-and-true-p`.  Perhaps something like this:

```lisp
(defun my/project-find-function (dir)
  (when-let ((match
              (and (bound-and-true-p eglot-lsp-context)
                   (locate-dominating-file dir "some-marker-file"))))
    `(transient . ,match)))

(add-hook 'project-find-functions #'my/project-find-function)
```

**Reply by cowboyd** [reply]
@joaotavora I tried a similar strategy to the one you outline, but I could not get it to work when switching to a project as opposed to just opening the file directly. My custom find function was never invoked by `(project-current)`. Unfortunately, finding a sub-package file is a very common use-case.

**Reply by joaotavora** [reply]
> Unfortunately, finding a sub-package file is a very common use-case.

I know, and sometimes you don't even have the luxury of a file marker.

Anyway, this is a bug in current Eglot and a slightly thinko when I developed this feature (and a mistest).  `bound-and-true-p` doesn't
return t when the symbol isn't dynamically bound, even if it is lexically bound.

A simple solution is to put `(setq eglot-lsp-context nil)` somewhere in your config.

**Reply by joaotavora** [reply]
Anyway, let's continue in the related and superbly notably named 1337 (lucky!) discussion, #1337
